### ISSUE TITLE
Suggestion - use generator yield for dispatching messages

### ISSUE THREAD

**Issue by samfrances** [ncthbrt/nact#52]
This framework looks extremely interesting.

I was wondering, have you considered following a "redux-saga" like model for dispatching messages between actors? e.g.:

```
const pingBehavior = async (msg, ctx) =>  {
  console.log(msg);
  await delay(500);
  yield dispatch(ctx.sender, ctx.name, ctx.self); // <--- here's where this differs from the docs example.
};

const ping = spawnStateless(system, pingBehavior, 'ping');
```

Where `dispatch()` would just produce an object which is used by the framework to do the actual sending:

```
const dispatch = (actor, msg, sender) => ({ DISPATCH: { actor, msg, sender }});
```

This would make actor "behaviour" functions potentially side-effect free, which would have testability benefits.

**Comment by samfrances** [comment]
[Redux Saga](https://redux-saga.js.org/) is probably a good place to look if you're looking for a model of how this sort of thing can work. Although, I wasn't necessarily suggesting a whole suite of declarative effects, just using this model for things like message dispatch.

**Comment by ncthbrt** [comment]
It definitely could be possible to perhaps have a parallel API for this purpose.

**Comment by ncthbrt** [comment]
@samfrances I'm quickly throwing together a prototype. Would appreciate a review when of that before proceeding to do the hard work of getting full coverage.

**Comment by ncthbrt** [comment]
@samfrances I've created a working prototype here: https://github.com/ncthbrt/nact/tree/generators
I've put an example in the examples folder.

**Comment by samfrances** [comment]
@ncthbrt Sure thing, I'll take a look tonight. Apologies, I didn't see your message until now.

**Comment by samfrances** [comment]
I don't know the `nact` codebase well enough yet to give well informed comments on the whole branch. However, from the examples, the API looks a lot like what I had in mind. I like how you have incorporated general IO as well as actor-specific actions.

**Comment by ncthbrt** [comment]
Think for this change to be worthwhile, should probably think about adding a testkit to probe actors and make assertions about their behaviour.

**Comment by biancode** [comment]
Is it in master, please? I got Cannot find module 'nact/monad' - require('nact/monad')

**Comment by ncthbrt** [comment]
@biancode It's not an officially supported yet (there is no documentation for this feature, less extensive tests), but the master branch of https://github.com/ncthbrt/nact has the monad variant of the library in it. You can install it by calling I believe `npm install https://github.com/ncthbrt/nact.git`

**Comment by ncthbrt** [comment]
I've released the monadic implementation of nact in `v7.2.2-beta0` of nact on npm, would love to get some feedback on it. (Note that I have not yet added typescript bindings for the monadic wrapper around nact)
